src/main/java/net/minecraftforge/common/brewing/ContainerBrewingRecipe.java

```diff
@@ -0,0 +1,122 @@
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.common.brewing;
+
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParseException;
+import net.minecraft.network.FriendlyByteBuf;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.util.GsonHelper;
+import net.minecraft.world.item.Item;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.alchemy.PotionUtils;
+import net.minecraft.world.item.crafting.Ingredient;
+import net.minecraft.world.item.crafting.RecipeSerializer;
+import net.minecraft.world.level.Level;
+import net.minecraftforge.common.ForgeMod;
+import net.minecraftforge.registries.ForgeRegistries;
+import net.minecraftforge.registries.ForgeRegistryEntry;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.function.Predicate;
+
+public record ContainerBrewingRecipe(@Nonnull ResourceLocation id,
+                                     @Nonnull Item base,
+                                     @Nonnull Ingredient reagent,
+                                     @Nonnull Item result) implements IBrewingRecipe
+{
+    @Override
+    public boolean matches(final BrewingContainerWrapper container, final Level level)
+    {
+        return getBase().test(container.base()) && getReagent().test(container.reagent());
+    }
+
+    @Override
+    public ItemStack assemble(final BrewingContainerWrapper container)
+    {
+        return PotionUtils.setPotion(new ItemStack(result()), PotionUtils.getPotion(container.base()));
+    }
+
+    @Override
+    public ItemStack getResultItem()
+    {
+        return ItemStack.EMPTY;
+    }
+
+    @Override
+    public ResourceLocation getId()
+    {
+        return id();
+    }
+
+    @Override
+    public RecipeSerializer<?> getSerializer()
+    {
+        return ForgeMod.CONTAINER_BREWING_SERIALIZER.get();
+    }
+
+    @Override
+    public Ingredient getReagent()
+    {
+        return reagent();
+    }
+
+    @Override
+    public Ingredient getBase()
+    {
+        return Ingredient.of(base());
+    }
+
+    public static class Serializer extends ForgeRegistryEntry<RecipeSerializer<?>> implements RecipeSerializer<ContainerBrewingRecipe>
+    {
+
+        @Override
+        public ContainerBrewingRecipe fromJson(final ResourceLocation p_199425_1_, final JsonObject p_199425_2_)
+        {
+            Item input = ForgeRegistries.ITEMS.getValue(new ResourceLocation(GsonHelper.getAsString(p_199425_2_, "base")));
```

Comment:
Could use `getValueOptional` and then `getOrThrow`. Seems odd to have the checks after everything has been deserialized.